server/services/v1/query_runner.go

```diff
@@ -430,78 +459,189 @@ type StreamingQueryRunner struct {
 	streaming Streaming
 }
 
-// Run is responsible for running/executing the query
-func (runner *StreamingQueryRunner) Run(ctx context.Context, tx transaction.Tx, tenant *metadata.Tenant) (*Response, context.Context, error) {
-	db, err := runner.GetDatabase(ctx, tx, tenant, runner.req.GetDb())
+type readerOptions struct {
+	from          keys.Key
+	ikeys         []keys.Key
+	table         []byte
+	noFilter      bool
+	inmemoryStore bool
+	filter        *filter.WrappedFilter
+	fieldFactory  *read.FieldFactory
+}
+
+func (runner *StreamingQueryRunner) buildReaderOptions(tenant *metadata.Tenant, db *metadata.Database, collection *schema.DefaultCollection) (readerOptions, error) {
+	var err error
+	var options = readerOptions{}
+	if options.filter, err = filter.NewFactory(collection.QueryableFields).WrappedFilter(runner.req.Filter); err != nil {
+		return options, err
+	}
+	if options.table, err = runner.encoder.EncodeTableName(tenant.GetNamespace(), db, collection); err != nil {
+		return options, err
+	}
+	if options.fieldFactory, err = read.BuildFields(runner.req.GetFields()); err != nil {
+		return options, err
+	}
+	if runner.req.Options != nil && len(runner.req.Options.Offset) > 0 {
+		if options.from, err = keys.NewKeyFromFDB(options.table, runner.req.Options.Offset); err != nil {
```

Comment:
We should make this generic and move FDB specifics inside KV package.
Maybe call this `keys.Deserialize`? or `keys.FromBinary`?

Reply:
Also I think it's better to serialize our key and return it as an offset to the user, instead of binary KV backend key.

Reply:
Regarding first, then it means the building of the key is at two places in the keys package and in the kv package. 

For (2), how will you be serializing our key? It is an array of interfaces. The binary backend key is the serialized form of our key. To serialize our key, we again need to iterate our primary keys defined in the schema otherwise there is no easy way to do it.